Add multi-threaded kahan sum aggregation. Optimize aggregation for different datatypes

kev-inn opened the pull request:
This commit improves the sum operation to use multiple threads. This is as fast as the current matrix implementation for all datatypes except String, which is slower and it should be discussed how a `sum` on Strings should even work, and Boolean, where it is a lot faster. If the matrix is empty though, it optimizes it to be a metadata operation (with non zero count), while the tensor operation does not yet support this.

Reshape instruction should also be as fast as matrix reshape and makes use of copy on write.

Note that all instructions are only implemented for CP-mode.

kev-inn committed:
Add multi-threaded kahan sum aggregation. Optimize aggregation for different datatypes.

This commit improves the sum operation to use multiple threads. This is as fast as the current matrix implementation for all datatypes except String, which is slower and it should be discussed how a `sum` on Strings should even work, and Boolean, where it is a lot faster. If the matrix is empty though it optimizes it to be a metadata operation (with non zero count), while the tensor does not yet support this.

Reshape instruction should also be as fast as matrix reshape and makes use of copy on write.

Note that all instructions are only implemented for CP-mode.

kev-inn commented:
Concerning the non optimized empty aggregation, I figured out the problem, the variable holding the tensor somehow is not put into the `LocalVariableMap`. I have yet to figure out why this is the case and where the code handling this for matrices is located.

kev-inn committed:
Merge branch 'master' into tensor_perf

kev-inn committed:
Fix merge. Add tensor to `LocalVariableMap`

kev-inn committed:
Add optimization of aggregation over empty tensor

mboehm7 commented:
so you didn't try out INT32 and FP32 yet, right? For these data types, we should also be about 2x faster.

kev-inn commented:
No, I did not, since I do not know how I can instantiate a literal of that type in DML, I did not find anything in that respect in the SystemML documentation. Is there a way in SystemDS?
Therefore I did also not optimize for those cases so it could be slow.

mboehm7 commented:
I would recommend, to just temporarily cast the double input tensor to the respective type inside your sum implementation and measure just the time for executing the sum (without the conversion).

mboehm7 commented on src/main/java/org/tugraz/sysds/runtime/data/DataTensorBlock.java:
```diff
@@ -1,23 +0,0 @@
-
-package org.tugraz.sysds.runtime.data;
-
-public class DataTensorBlock extends TensorBlock
```
Please keep this as a reminder of implementing DataTensors, which can be composed of tensor blocks of certain types plus a column mapping. I'd like to early on decide if we should implement operations in TensorBlock and then composite operations in DataTensorBlock or if we should directly focus on DataTensorBlock.

kev-inn replied:
Sorry, of course. I deleted this file on accident, because I mixed it up with my temporary Data extended TensorBlock

mboehm7 commented on src/main/java/org/tugraz/sysds/runtime/data/DenseBlockBool.java:
```diff
@@ -44,12 +44,13 @@ protected void allocateBlock(int bix, int length) {
 		_data = new BitSet(length);
 	}
 
-	public DenseBlockBool(int[] dims, boolean[] data) {
```
Could we please keep both constructors, as passing a boolean[] array might be useful for conversion purposes.

mboehm7 commented on src/main/java/org/tugraz/sysds/runtime/data/LibTensorAgg.java:
```diff
@@ -0,0 +1,396 @@
+		final int n = in.getDim(1);
+		final int cix = (m-1)*n;
+
+		double[] a = in.getDenseBlock().valuesAt(0);
```
Please add a TODO for generalize for large blocks if you use it for local operations of large matrices (not vectors).

mboehm7 commented on src/main/java/org/tugraz/sysds/runtime/data/LibTensorAgg.java:
```diff
@@ -0,0 +1,396 @@
+				int lpos = (bix == bil) ? a.pos(rl) : 0;
+				int len = (bix == biu) ? a.pos(ru - 1) - lpos + a.getCumODims(0) : a.blockSize(bix) * a.getCumODims(0);
+				for (int i = 0; i < len; i++)
+					sum += UtilFunctions.toLong(a.get(bix * a.blockSize() * a.getCumODims(0), lpos + i));
```
Did you try to iterate in two loops, rows and other dimensions (from cumODims) and doing a simple get(i,j)? Also, why is there a toLong cast in the case that also applies to FP32?

kev-inn replied:
with the simple get(i,j) are you refering to the FP64 case?
My mistake, I should of course NOT cast toLong.

mboehm7 replied:
I was thinking of something like https://github.com/tugraz-isds/systemds/blob/master/src/main/java/org/tugraz/sysds/runtime/data/TensorBlock.java#L371 (except that we don't even need a switch if we simply compute a double sum) as a fallback for all data types (except string) where we do not want to integrate a custom implementation.

mboehm7 reviewed (commented):
Overall, this is an excellent start @kev-inn - let's do some cleanup and then we can merge it in. Before we go further ahead with these operations though, we should have our next call to make sure we're all happy with the general template. It would be awesome to find a clean, understandable, general-case implementation (for all remaining data types) that closely matches the expected performance for memory-bandwidth bound operations.

kev-inn committed:
Restore deleted code and improve `uakp` sum loop

kev-inn committed:
Merge branch 'master' into tensor_perf

kev-inn commented:
I fixed the issues, but I did not yet do in depth performance tests for int32 or float. The tests I did showed that performance is similar to long and double, so there should be space for improvement. I will think about how we can make the current implementation a bit cleaner, while not sacrificing performance, so we have a nice template for future implementations.

kev-inn committed:
Made code more compact

This commit tries to simplify the code a bit, this leads to some differences compared to the matrix sum aggregation.

I did some performance tests for a specialized sum implementation for FP64 (making use of `DenseBlock.valuesAt()`), but got worse performance for a variety of block sizes (100kx1k, 10kx1k, 1kx1k, 100x1k and 10x1k), so I will keep the general implementation for now and we can have another look into it later.

mboehm7 commented:
LGTM - thanks @kev-inn - I only made minor changes to remove unnecessary warnings and fix an issue in `TensorBlock.copy()`
